File: Algorithms/C/BinarySearch/binarySearch.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Recursive Method

int search(int array[], int x, int low, int high) {
  if (high >= low) {
    int mid = low + (high - low) / 2;

    // If found at mid, then return it
    if (array[mid] == x)
      return mid;

    // Search the left half
    if (array[mid] > x)
      return search(array, x, low, mid - 1);

    // Search the right half
    return search(array, x, mid + 1, high);
  }

  return -1;
}

int binarySearch(int arraySize)
{

    int array[arraySize];
    int maxNum = arraySize;

    srand(time(0));

    for (int i = 0; i < arraySize; i++)
    {
        array[i] = rand() % maxNum;
    }

    int x = rand() % maxNum;

    int result = search(array, x, 0, arraySize - 1);

    // for (int i = 0; i < arraySize; i++)
    //     printf(" %d \n", array[i]);

    if (result == -1)
    {
        printf("%d Not found \n", x);
    }
    else {
        printf("%d Found at index %d \n", x, result);
    }

    return 0;
}

File: Algorithms/Zig/BinarySearch/binarySearch.zig
const std = @import("std");

extern fn print(u32) void;

fn search (myArray: []u8, x: u8, low: i8, high: i8) i8 {
    while (low <= high)
    {
        var mid: i32 = low + (high - low) / 2;

        if (myArray[mid] == x) {
            return mid;
        }

        if (myArray[mid] < x) {
            low = mid + 1;
        }
        else {
            high = mid - 1;
        }
    }

    return -1;
}


fn binarySearch() void {

    var array: [10]u8 = .{11, 22, 33, 44, 34, 77, 234, 63, 44, 2};
    var result: u8 = search(array, 11, 0, 9);

    if (result == -1)
    {
        print(search ++ " Not found \n");
    }
    else {
        print(search ++ " Found at index \n" ++ result);
    }

    return 0;
}

pub fn main() void {
    binarySearch();
}